<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<title>[Python Mocking] #python #mocking #mocks #tornado</title>
<link rel="stylesheet" type="text/css" href="../assets/css/styles.css">
</head>
<body>
<div class="container">
<a href="../index.html" class="back-link">&laquo; Back to Index</a>
<h1 id="python-mocking-python-mocking-mocks-tornado">[Python Mocking] #python #mocking #mocks #tornado</h1>

<h2 id="python-mocking-md">Python Mocking.md</h2>

<ul>
<li><a href="#unittestmock-or-mock">unittest.mock or mock</a></li>
<li><a href="#verify-exceptions">Verify Exceptions</a></li>
<li><a href="#clearing-lru_cache">Clearing lru_cache</a></li>
<li><a href="#mock-module-levelglobal-variables">Mock Module Level/Global Variables</a></li>
<li><a href="#mock-instance-method">Mock Instance Method</a></li>
<li><a href="#mock-class-method">Mock Class Method</a></li>
<li><a href="#mock-entire-class">Mock Entire Class</a></li>
<li><a href="#mock-async-calls">Mock Async Calls</a></li>
<li><a href="#mock-instance-types">Mock Instance Types</a></li>
<li><a href="#mock-return_value-vs-side_effect">Mock return_value vs side_effect</a></li>
<li><a href="#mock-nested-calls">Mock Nested Calls</a></li>
<li><a href="#mock-builtin-open-function">Mock builtin <code>open</code> function</a></li>
</ul>

<h2 id="unittest-mock-or-mock">unittest.mock or mock</h2>

<p>The <code>mock</code> module is a backwards compatible library you can download from PyPy, where as <code>unittest.mock</code> is the same thing but only compatible with the version of Python you&rsquo;re using.</p>

<p>So in almost all cases you&rsquo;ll want to import it like so:</p>

<pre><code class="language-py">import unittest.mock as mock
</code></pre>

<blockquote>
<p>For more examples, see <a href="http://www.voidspace.org.uk/python/mock/examples.html" target="_blank">this reference guide</a></p>
</blockquote>

<h2 id="verify-exceptions">Verify Exceptions</h2>

<p>In the following example our code (in <code>app.account.confirm(...)</code>) catches a generic <code>Exception</code> and re-raises it as <code>exceptions.CognitoException</code>, which we catch and make assertions against:</p>

<pre><code class="language-py">@mock.patch('app.aws.sdk.confirm_sign_up', side_effect=Exception('whoops'))
@mock.patch('app.account.instr_exc')
def test_account_confirm_failure(mock_instr_exc, mock_signup):
    with pytest.raises(exceptions.CognitoException) as exc_info:
        app.account.confirm(123, 'foo')
        assert True is True  # this will never be executed!
        
    assert exc_info.typename == 'CognitoException'
    assert str(exc_info.value) == 'SIGNUP_CONFIRMATION_FAILED'

    # we can't have a single assert against call_args_list because in python
    # two instances of an exception aren't considered equal, and so we have to
    # assert individual elements of call_args_list.
    assert type(mock_instr_exc.call_args_list[0][0][0]) == Exception
    assert str(mock_instr_exc.call_args_list[0][0][0]) == 'whoops'
    assert mock_instr_exc.call_args_list[0][0][1] == 'SIGNUP_CONFIRMATION_FAILED'
    assert mock_instr_exc.call_args_list[0][1] == {'code': '123'}
</code></pre>

<blockquote>
<p>Note: don&rsquo;t make the mistake of putting any assertions within the <code>with</code> context manager. Once the Exception is raised by the function being called within the <code>with</code> context manager, all code after it inside the block is not executed.</p>
</blockquote>

<h2 id="clearing-lru-cache">Clearing lru_cache</h2>

<p>If a function you wish to test has the <code>functools.lru_cache</code> decorator applied, then you&rsquo;ll need to be mindful of mocking the response of that function as it&rsquo;ll be cached in one test and the cached result will be returned when calling the function again to test some other behaviour (and might likely confuse you when you see the unexpected response).</p>

<p>To fix this issue is very easy because <code>lru_cache</code> provides additional functions when decoratoring your functions, it provides:</p>

<ul>
<li><code>cache_info</code></li>
<li><code>cache_clear</code></li>
</ul>

<p>The latter (<code>cache_clear</code>) is what you call&hellip;</p>

<pre><code class="language-py">@lru_cache(5)
def foo():
    print('Executing foo...')
    
foo()  # Executing foo...
foo()  # &lt;nothing printed as None response was cached and returned&gt;
foo.cache_info()  # CacheInfo(hits=1, misses=1, maxsize=5, currsize=1)
foo.cache_clear()
foo()  # Executing foo... (notice the 'side effect of print is executed again)
</code></pre>

<blockquote>
<p>Note: debugging this isn&rsquo;t always obvious. Later on I demonstrate how to <a href="#mock-builtin-open-function">mock the builtin <code>open</code> function</a>, and in that scenario I stumbled across this issue, because although I wasn&rsquo;t mocking the top level function itself (I was mocking the call to <code>open</code> within), the contents of the file being opened was what was returned and being cached.</p>
</blockquote>

<h2 id="mock-module-level-global-variables">Mock Module Level/Global Variables</h2>

<p>With a module variable you can can either set the value directly or use <code>mock.patch</code>.</p>

<p>In the following example we have the variable <code>client_id</code> which is a global variable inside the <code>app.aws</code> module which we import to reference elsewhere in our code:</p>

<pre><code class="language-py">import app.aws


def test_account_confirm_successful():
    app.aws.client_id = 456  # used internally by `confirm()`
    ...
    
@mock.patch('app.aws.client_id', 456)
def test_account_confirm_successful():
    ...
</code></pre>

<p>In the <code>mock.patch</code> example, there are two key things to notice:</p>

<ol>
<li>we don&rsquo;t use <code>return_value</code>.</li>
<li>there is no mock instance passed to the test function.</li>
</ol>

<h2 id="mock-instance-method">Mock Instance Method</h2>

<p>Mock the entire class and take advantage of the fact that a mock, when called, returns a new mock:</p>

<pre><code class="language-python">@mock.patch(&quot;foo.bar.SomeClass&quot;)
def test_stuff(mock_class):
    mock_class.return_value.made_up_function.return_value = &quot;123&quot;
</code></pre>

<p>The reason this ^^ works is because calling a mock <em>returns another mock</em>, and so if you call <code>mock_class.return_value</code> you&rsquo;re actually getting another mock object; and because you can call anything you like on a mock object (a function or property you call on a mock doesn&rsquo;t have to exist), means you can set a <code>return_value</code> on the mock that&rsquo;s returned by calling <code>made_up_function</code>.</p>

<h2 id="mock-class-method">Mock Class Method</h2>

<p>Similar approach to mocking an instance method in that you mock the entire class but you have one less <code>return_value</code> to assign to:</p>

<pre><code class="language-python">mock_class.ClassMethodName.return_value = &quot;123&quot;
</code></pre>

<h2 id="mock-entire-class">Mock Entire Class</h2>

<p>To mock an entire class you&rsquo;ll need to set the <code>return_value</code> to be a new instance of the class.</p>

<pre><code class="language-py">@mock.patch('myapp.app.Car')
def test_class(self, mock_car):

    class NewCar(object):
        def get_make(self):
            return 'Audi'

        @property
        def wheels(self):
            return 6

    mock_car.return_value = NewCar()
    ...
</code></pre>

<blockquote>
<p>See other class related mocking tips <a href="https://chase-seibert.github.io/blog/2015/06/25/python-mocking-cookbook.html" target="_blank">here</a></p>
</blockquote>

<h2 id="mock-async-calls">Mock Async Calls</h2>

<p>We can create a coroutine and allow it to be configurable for different types of responses:</p>

<pre><code class="language-python">def make_coroutine(response):
    &quot;&quot;&quot;You could pass response as a mock or as a raw data structure.&quot;&quot;&quot;
    
    async def coroutine(*args, **kwargs):
        &quot;&quot;&quot;Imagine this coroutine is called with a url as the first argument.&quot;&quot;&quot;
        
        if args[0] == '/exception/foo':
            raise Exception('Whoops Foo')
        elif args[0] == '/exception/bar':
            raise Exception('Whoops Bar')
        return response
    return coroutine
    
class TestTornado(AsyncHTTPSTestCase):
    def get_app(self):
        class FakeHandler(tornado.web.RequestHandler):
            async def get(self, *args, **kwargs):
                self.finish('hello')
                
        return tornado.web.Application([
            (r'/fake', FakeHandler),
        ])
    
    @mock.patch('foo.bar.http_client')
    def test_async thing(self, mock_client):
        response_body = {'state': 'success', 'payload': 'foobar'}
        response_mock = mock.MagicMock()
        response_mock.body = json.dumps(response_body)
        
        mock_client.post.side_effect = make_coroutine(response_mock)

        response = self.fetch('/fake')
        assert response.code == 200
        assert json.loads(response.body.decode()) == response_body
</code></pre>

<blockquote>
<p>Note: you could also create a <code>MagicMock</code> and set properties on it like <code>m = mock.MagicMock(x=1, y=2, z=3)</code> and then pass that into the <code>make_coroutine</code> function. That way, within each <code>if</code> statement you can then just call <code>m.x</code> or <code>m.y</code> etc to get at the actual response you want to return (rather than having to hardcode response objects within the function itself). Mock is also considered <code>callable</code> (see below implementation).</p>
</blockquote>

<p>When dealing with side_effects that need to sometimes trigger an Exception and other times suceed you could use a slightly modified mock implementation that checks if the given response object is callable or not&hellip;</p>

<pre><code class="language-python">count = 0

def make_side_effect_coroutine(side_effect):
    &quot;&quot;&quot;Side effect friendly mock coroutine.

    In some tests we need to have a mocked coroutine return a different value
    when it's called multiple times, but a mock side_effect can't trigger a
    raised exception when given an iterator, and so we have to construct that
    behaviour ourselves.
    &quot;&quot;&quot;

    async def coroutine(*args, **kwargs):
        return side_effect(*args, **kwargs) if callable(side_effect) else side_effect
    return coroutine
    
@mock.patch('app.thing')
def test_confirm_email_change_failure(self, mock_thing):

    def side_effects(*args, **kwargs):
        &quot;&quot;&quot;Use global var to control mock side effects.&quot;&quot;&quot;

        global count

        if count &gt; 0:
            raise Exception('whoops')

        count += 1
        return  # don't raise an exception the first time around

    mock_thing.side_effect = make_side_effect_coroutine(side_effects)
</code></pre>

<p>Alternatives&hellip;</p>

<h3 id="monkey-patch">Monkey Patch</h3>

<pre><code class="language-py"># allow mock to be used as an await expression...

async def async_response():
    return namedtuple('_', ['body'])('{&quot;state&quot;: &quot;success&quot;}')


def mock_async_expression(our_mock):
    return async_response().__await__()


mock.MagicMock.__await__ = mock_async_expression
</code></pre>

<h3 id="magicmock-subclass">MagicMock Subclass</h3>

<pre><code class="language-py">class AsyncMock(MagicMock):
    async def __call__(self, *args, **kwargs):
        return super(AsyncMock, self).__call__(*args, **kwargs)
        
class TestHandlers(testing.AsyncTestCase):
    @mock.patch('app.handlers.trigger_soft_cdn_purge', new_callable=AsyncMock)
    @mock.patch('app.handlers.api')
    @testing.gen_test
    async def test_update_cache(self, api_mock, trigger_soft_cdn_purge):
        response = mock.MagicMock()
        response.code = 200
        api_mock.buzz = AsyncMock(return_value=response)
</code></pre>

<h3 id="async-inline-function">Async Inline Function</h3>

<pre><code class="language-py">@mock.patch('app.buzz_api.api_gateway')
@testing.gen_test
async def test_buzz_api(self, client_mock):
    async def get(url, **kwargs):
        return
        
    client_mock.get.side_effect = get
</code></pre>

<h2 id="mock-instance-types">Mock Instance Types</h2>

<p>There are two ways to make a mock more like the real object being mocked.</p>

<ol>
<li><code>spec</code></li>
<li><code>wrap</code></li>
</ol>

<p>We can use mock&rsquo;s <code>spec</code> feature to mimic all methods/attributes of the object being mocked. This ensures your mocks have the same api as the objects they are replacing.</p>

<blockquote>
<p>Note: there is a stricter <code>spec_set</code> that will raise an <code>AttributeError</code>.</p>
</blockquote>

<p>This is best demonstrated with an example:</p>

<pre><code class="language-py">import unittest.mock as mock
import tornado.simple_httpclient

from tornado.httpclient import AsyncHTTPClient


http_client = AsyncHTTPClient()
type(http_client)  # tornado.simple_httpclient.SimpleAsyncHTTPClient

isinstance(http_client, tornado.simple_httpclient.SimpleAsyncHTTPClient)  # True

isinstance(mock.MagicMock(), tornado.simple_httpclient.SimpleAsyncHTTPClient)  # False

m = mock.MagicMock(spec=tornado.simple_httpclient.SimpleAsyncHTTPClient)
isinstance(m, tornado.simple_httpclient.SimpleAsyncHTTPClient)  # True
</code></pre>

<p>The <code>wrap</code> parameter allows you to &lsquo;spy&rsquo; on the implementation, as wll as affect it&rsquo;s behaviour:</p>

<pre><code class="language-py">@pytest.mark.parametrize(&quot;input_date, input_url, valid&quot;, [
    (&quot;2017-06-17T00:00:00.000000Z&quot;, &quot;foo&quot;, True),
    (&quot;2017-06-18T00:00:00.000000Z&quot;, &quot;bar&quot;, False),
])
@mock.patch(&quot;app.handlers.data.datetime&quot;, wraps=datetime)
def test_valid_video(mock_datetime, input_date, input_url, valid):
    mock_datetime.now.return_value = datetime(2017, 6, 18, 00, 00, 00, 000000)
    assert valid_video(input_date, input_url) is valid
</code></pre>

<h2 id="mock-return-value-vs-side-effect">Mock return_value vs side_effect</h2>

<p>If your function has a try/except around it, then you can use <code>side_effect</code> to cause the calling of the function to trigger an Exception as the returned value:</p>

<pre><code class="language-py">@mock.patch('app.aws.sdk.confirm_sign_up', side_effect=Exception('whoops'))
</code></pre>

<blockquote>
<p>Note: if you had used <code>return_value=Exception('whoops')</code> then the mock would return the string representation of the Exception rather than <em>raising</em> an exception like <code>side_effect</code> does.</p>
</blockquote>

<p>Otherwise if you just need a <em>static</em> value returned, so it&rsquo;s evaluated at the time it&rsquo;s defined (not when it&rsquo;s called), then you can use <code>return_value</code> instead:</p>

<pre><code class="language-py">@mock.patch('app.security.secret_hash', return_value='###')
</code></pre>

<h2 id="mock-nested-calls">Mock Nested Calls</h2>

<p>Calling a property on a mock returns another mock, so in order to mock very specific properties you&rsquo;ll need to nest your <code>return_value</code> or <code>side_effect</code>:</p>

<pre><code class="language-py">m = mock.MagicMock()
m.return_value.get.side_effect = [1, 2]
m.return_value.post.return_value = 'foo'

x = m()

x.get()   # 1
x.post()  # foo
x.get()   # 2
</code></pre>

<h2 id="mock-builtin-open-function">Mock builtin <code>open</code> function</h2>

<p>Python&rsquo;s mock library provides an abstraction for mocking the builtin <code>open</code> function a lot simpler&hellip;</p>

<pre><code class="language-py">def test_load_ui_messages_successful():
    &quot;&quot;&quot;Verify ui message YAML file can be read properly.&quot;&quot;&quot;

    file_content = 'foo: bar'

    with mock.patch('bf_auth.utility.open', mock.mock_open(read_data=file_content), create=True) as mock_builtin_open:
        assert utils.load_ui_messages('./path/to/non/existing/file.yaml') == {'foo': 'bar'}
</code></pre>

<p>The <code>create=True</code> param set on <code>mock.patch</code> means that the <code>mock.MagicMock</code> returned will automagically create any attributes that are called on the mock (this is because the <code>open</code> function will attempt to access lots of different things and it&rsquo;s easier for mock to mock out all of that for you).</p>

</div>
</body>
</html>